lib/os/mpsc_pbuf.c

```diff
@@ -0,0 +1,399 @@
+				uint32_t idx, uint32_t val)
+{
+	uint32_t i = idx - val;
+
+	if (buffer->flags & MPSC_PBUF_SIZE_POW2) {
+		return idx & (buffer->size - 1);
+	}
+
+	return (i >= buffer->size) ? i + buffer->size : i;
+}
+
+static inline uint32_t get_skip(union mpsc_pbuf_generic *item)
+{
+	if (item->hdr.busy && !item->hdr.valid) {
+		return item->skip.len;
+	}
+
+	return 0;
+}
+
+static void add_skip_item(struct mpsc_pbuf_buffer *buffer, uint32_t wlen)
+{
+	union mpsc_pbuf_generic skip = {
+		.skip = { .valid = 0, .busy = 1, .len = wlen }
+	};
+
+	buffer->buf[buffer->tmp_wr_idx] = skip.raw;
+	buffer->tmp_wr_idx = idx_inc(buffer, buffer->tmp_wr_idx, wlen);
+	buffer->wr_idx = idx_inc(buffer, buffer->wr_idx, wlen);
+}
+
+static union mpsc_pbuf_generic *drop_item_locked(struct mpsc_pbuf_buffer *buffer,
+						 uint32_t free_wlen)
+{
+	union mpsc_pbuf_generic *item;
+	uint32_t rd_wlen;
+	uint32_t skip_wlen;
+
+	item = (union mpsc_pbuf_generic *)&buffer->buf[buffer->rd_idx];
+	skip_wlen = get_skip(item);
+
+	rd_wlen = skip_wlen ? skip_wlen : buffer->get_wlen(item);
+	if (skip_wlen) {
+		item = NULL;
+	} else if (item->hdr.busy) {
+		/* item is currently processed and cannot be overwritten. */
+		add_skip_item(buffer, free_wlen + 1);
+		buffer->wr_idx = idx_inc(buffer, buffer->wr_idx, rd_wlen);
+		buffer->tmp_wr_idx = idx_inc(buffer, buffer->tmp_wr_idx, rd_wlen);
+
+		/* Get next itme followed the busy one. */
+		uint32_t next_rd_idx = idx_inc(buffer, buffer->rd_idx, rd_wlen);
+
+		item = (union mpsc_pbuf_generic *)&buffer->buf[next_rd_idx];
+		skip_wlen = get_skip(item);
+		if (skip_wlen) {
+			item = NULL;
+			rd_wlen += skip_wlen;
+		} else {
+			rd_wlen += buffer->get_wlen(item);
+		}
+
+	}
+
+	buffer->rd_idx = idx_inc(buffer, buffer->rd_idx, rd_wlen);
+	buffer->tmp_rd_idx = buffer->rd_idx;
+
+	return item;
+}
+
+void mpsc_pbuf_put_word(struct mpsc_pbuf_buffer *buffer,
+			union mpsc_pbuf_generic item)
+{
+	bool cont;
+	uint32_t free_wlen;
+	k_spinlock_key_t key;
+	union mpsc_pbuf_generic *dropped_item = NULL;
+
+	do {
+
+		cont = false;
+		key = k_spin_lock(&buffer->lock);
+		(void)free_space(buffer, &free_wlen);
+		if (free_wlen) {
+			buffer->buf[buffer->tmp_wr_idx] = item.raw;
+			buffer->tmp_wr_idx = idx_inc(buffer,
+						     buffer->tmp_wr_idx, 1);
+			buffer->wr_idx = idx_inc(buffer, buffer->wr_idx, 1);
+		} else if (buffer->flags & MPSC_PBUF_MODE_OVERWRITE) {
+			dropped_item = drop_item_locked(buffer, free_wlen);
+			cont = true;
+		} else {
+			/* empty */
+		}
+
+		k_spin_unlock(&buffer->lock, key);
+
+		if (cont && dropped_item) {
+			/* Notify about item being dropped. */
+			buffer->notify_drop(buffer, dropped_item);
+		}
+	} while (cont);
+
+}
+
+union mpsc_pbuf_generic *mpsc_pbuf_alloc(struct mpsc_pbuf_buffer *buffer,
+					 size_t wlen, k_timeout_t timeout)
+{
+	union mpsc_pbuf_generic *item = NULL;
+	union mpsc_pbuf_generic *dropped_item = NULL;
+	bool cont;
+	uint32_t free_wlen;
+
+	do {
+		cont = false;
```

Comment:
better would be to put the cont after the variables are introduced (so three lines after this one).